inc/aes_packages/namespace.php

```diff
@@ -0,0 +1,318 @@
+	if ( $for_network ) {
+		$existing_package_id = get_site_option( "altis_search_package_{$slug}", false );
+	} else {
+		$existing_package_id = get_option( "altis_search_package_{$slug}", false );
+	}
+	if ( $existing_package_id ) {
+		// The cleanup does not have to happen straight away so postpone it for an hour
+		// to let the settings updates and data indexing complete.
+		wp_schedule_single_event( time() + HOUR_IN_SECONDS, 'altis.search.dissociate_package', [
+			$existing_package_id,
+		] );
+	}
+
+	try {
+		// Derive S3 bucket and path.
+		preg_match( '#^s3://([^/]+)/(.*?)$#', $file, $s3_file_parts );
+		$s3_bucket = $s3_file_parts[1];
+		$s3_key = $s3_file_parts[2];
+
+		// Create a new package.
+		$new_package = $client->createPackage( [
+			'PackageName' => $name, // required.
+			'PackageDescription' => $file,
+			'PackageSource' => [ // required.
+				'S3BucketName' => $s3_bucket,
+				'S3Key' => $s3_key,
+			],
+			'PackageType' => 'TXT-DICTIONARY',
+		] );
+
+		// Association can fail the first time if we run it too quickly.
+		usleep( 100000 );
+
+		// Associate the package with the ES domain.
+		// NOTE: This process is async so cannot be applied immediately.
+		$client->associatePackage( [
+			'DomainName' => $domain, // required.
+			'PackageID' => $new_package['PackageDetails']['PackageID'], // required.
+		] );
+
+		// Set package status.
+		if ( $for_network ) {
+			update_site_option( "altis_search_package_status_{$slug}", 'ASSOCIATING' );
+		} else {
+			update_option( "altis_search_package_status_{$slug}", 'ASSOCIATING' );
+		}
+	} catch ( Exception $e ) {
+		// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+		trigger_error( 'Error creating package on AES: ' . $e->getMessage(), E_USER_WARNING );
+		return null;
+	}
+
+	$package_id = sprintf( 'analyzers/%s', $new_package['PackageDetails']['PackageID'] );
+
+	// Return the AES package ID string.
+	return $package_id;
+}
+
+/**
+ * Check whether the package has finished associating with the AES domain
+ * before returning it.
+ *
+ * @param string|null $package_id The stored package ID.
+ * @param string $slug The slug for the package.
+ * @param boolean $for_network Whether this is a network level package.
+ * @return string|null
+ */
+function get_package_id( ?string $package_id, string $slug, bool $for_network = false ) {
+	// Check status.
+	if ( $for_network ) {
+		$status = get_site_option( "altis_search_package_status_{$slug}", 'ASSOCIATING' );
+	} else {
+		$status = get_option( "altis_search_package_status_{$slug}", 'ASSOCIATING' );
+	}
+
+	// Queue up a background task if this is still pending.
+	// Association usually takes 1-5 minutes so use a 30 second resolution.
+	$hook_args = [ $package_id, $slug, $for_network ];
+	if ( $package_id && $status !== 'ACTIVE' && ! wp_next_scheduled( 'altis.search.check_package_status', $hook_args ) ) {
+		wp_schedule_single_event( time() + 60, 'altis.search.check_package_status', $hook_args );
```

Comment:
This seems like an odd side-effect for a `get_` function to do, could this be moved to the caller?

Reply:
Not easily. It's a bit of an odd one. This means it'll only do the lookup when the data is requested but I'll see if I can make the chain of background tasks more sensible.